ethereum/contracts/ERC721App.sol

```diff
@@ -0,0 +1,142 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.5;
+pragma experimental ABIEncoderV2;
+
+import "@openzeppelin/contracts/access/AccessControl.sol";
+import "@openzeppelin/contracts/token/ERC721/extensions/IERC721Metadata.sol";
+import "./ScaleCodec.sol";
+import "./OutboundChannel.sol";
+
+enum ChannelId {
+    Basic,
+    Incentivized
+}
+
+contract ERC721App is AccessControl {
+    using ScaleCodec for uint256;
+
+    mapping(ChannelId => Channel) public channels;
+
+    bytes2 constant MINT_CALL = 0x4301;
+
+    event Locked(
+        address tokenContract,
+        uint256 tokenId,
+        address sender,
+        bytes32 recipient
+    );
+
+    event Unlocked(
+        address tokenContract,
+        uint256 tokenId,
+        bytes32 sender,
+        address recipient
+    );
+
+    struct Channel {
+        address inbound;
+        address outbound;
+    }
+
+    bytes32 public constant INBOUND_CHANNEL_ROLE =
+        keccak256("INBOUND_CHANNEL_ROLE");
+
+    constructor(Channel memory _basic, Channel memory _incentivized) {
+        Channel storage c1 = channels[ChannelId.Basic];
+        c1.inbound = _basic.inbound;
+        c1.outbound = _basic.outbound;
+
+        Channel storage c2 = channels[ChannelId.Incentivized];
+        c2.inbound = _incentivized.inbound;
+        c2.outbound = _incentivized.outbound;
+
+        _setupRole(INBOUND_CHANNEL_ROLE, _basic.inbound);
+        _setupRole(INBOUND_CHANNEL_ROLE, _incentivized.inbound);
+    }
+
+    /**
+     * @notice Locks ERC721 token in ERC721App contract
+     * @param _tokenContract The NFT contract
+     * @param _tokenId The NFT to lock
+     * @param _recipient Polkadot address of the receiver
+     * @param _channelId The Channel to use to send token
+     */
+    function lock(
+        address _tokenContract,
+        uint256 _tokenId,
+        bytes32 _recipient,
+        ChannelId _channelId
+    ) public payable {
+        require(
+            _channelId == ChannelId.Basic ||
+                _channelId == ChannelId.Incentivized,
+            "Invalid channel ID"
+        );
+
+        IERC721Metadata token = IERC721Metadata(_tokenContract);
+
+        token.transferFrom(msg.sender, address(this), _tokenId);
```

Comment:
@philipstanislaus is there re-entrancy risk here? (eg: malicious erc20 that does weird shit in it's transferFrom)

Reply:
I'm guessing not, because even with re-entrancy I don't think anything can be stolen/fucked up? But should review in more detail.

Reply:
As before, this should be fine, since there is no state read/written before the transferFrom call. Having said that, it is recommended to first finish all state changes first, before calling the external function.

Reply:
will fix